stats/Record.md

```diff
@@ -0,0 +1,73 @@
+# Record API Overview
+The stats library MUST allow users to record metrics for their applications or third-party
+libraries. The core data types used are:
+* `Measure`: describes the type of the individual values recorded by an application.
+* `Measurement`: describes a data point to be collected for a `Measure`.
+
+## Measure
+A `Measure` describes the type of the individual values/measurements recorded by an application. It
+includes information such as the type of measurement, the units of measurement and descriptive names
+for the data. This provides the fundamental type used for recording data.
+
+A Measure describes a value with the following metadata:
+* `name`: a string by which the measure will be referred to, e.g. "rpc_server_latency", or
+"vm_cpu_cycles". Names MUST be unique within the library. It is recommended to use names compatible
+with the intended end usage, e.g, use host/path pattern.
+* `description`: a free format descriptive string, e.g. "RPC latency in seconds", "Virtual cycles
+executed on VM".
+* `unit`: a string describing the unit's used for this Measure. Follows the format described by
+[Unified Code for Units of Measure](http://unitsofmeasure.org/ucum.html).
+* `type`: the only supported types are `int64` and `double`.
+
+Implementations MUST define a Measure class, constructed from the parameters above. Measure MAY have
+getters for retrieving all of the information used in `Measure` definition. Once created, Measure
+metadata is immutable.
+
+Example in Java
+```
+private static final MeasureDouble rpcLatency =
+    MeasureDouble.create("grpc.io/latency", "latency", "ms");
+```
+
+References to Measures in the system MAY be obtained from querying of registered Measures. This
+functionality is required to decouple the recording of the data from the exporting of the data.
+
+Example in Java
+```
+private static final MeasureDouble rpcLatency = Stats.findMeasureDoubleByName("grpc.io/latency");
+```
+
+Implementations MAY define a `MeasureDescription` class which contains all the read-only fields from
+the `Measure` definition such as: `name`, `description`, `unit` and `type`.
+
+## Measurement / MeasurementMap
+A `Measurement` describes a value with the following metadata:
+* `Measure`: the `Measure` to which this `value` is applied.
+* `value`: recorded value, MUST have the appropriate type to match the `Measure` definition.
+
+Implementations MAY define a `MeasurementMap` which describes a set of data points to be collected
+for a set of Measures. Adding this functionality may improve the record usage API.
+
+## Record usage
+
+Users MUST record Measurements with the current context (implicit or explicit). Tags from the
+current context are recorded with the Measurements if they are any.
+
+Implementations MUST provide a means of recording Measurements. This functionality SHOULD be
+provided through one of the following options:
```

Comment:
We need to encourage recording multiple data points at a single time to reduce the deserialization cost.

Shouldn't we require the first option as a MUST and the second one optionally?

Reply:
The second is still batch. MeasurementMap is essentially a collection of Measurements. See the example in Java.